### ISSUE TITLE
vmware_deploy_ovf is unable to attach NSX based PortGroup

### ISSUE THREAD

**Issue by Udayendu** [ansible/ansible#63376]
##### SUMMARY
vmware_deploy_ovf is unable to attach NSX based PortGroup

Seems like this feature is available in vmware_guest ansible module: https://github.com/ansible/ansible/pull/37979

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
vmware_deploy_ovf

##### ANSIBLE VERSION
```
ansible --version
ansible 2.9.0rc1
  config file = None
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.6/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.6.8 (default, Aug 20 2019, 17:12:48) [GCC 8.3.0]
```

##### CONFIGURATION
```
a. ESXi 6.5 or 6.7
b. vCenter 6.5 or 6.7
```

##### OS / ENVIRONMENT
Ubuntu 18.04. LTS

$ pip3 show pyvmomi
Name: pyvmomi
Version: 6.7.3
Summary: VMware vSphere Python SDK
Home-page: https://github.com/vmware/pyvmomi
Author: VMware, Inc.
Author-email: jhu@vmware.com
License: License :: OSI Approved :: Apache Software License
Location: /usr/local/lib/python3.6/dist-packages
Requires: six, requests

$  pip3 show ansible
Name: ansible
Version: 2.9.0rc1
Summary: Radically simple IT automation
Home-page: https://ansible.com/
Author: Ansible, Inc.
Author-email: info@ansible.com
License: GPLv3+
Location: /usr/local/lib/python3.6/dist-packages
Requires: PyYAML, jinja2, cryptography

##### STEPS TO REPRODUCE

```
# cat deploy-ova.yml
---
- name: PaloAlto Network Firewall 8.0.5 Deployment
  hosts: localhost
  connection: local
  gather_facts: false
  vars_files:
    - vsphere_creds.yml
    - vsphere_var.yml
  tasks:
    - name: VM Deployment from OVA
      vmware_deploy_ovf:
        hostname: '{{ vcenter_hostname }}'
        username: '{{ vcenter_username }}'
        password: '{{ vcenter_password }}'
        datacenter: '{{ vsphere_datacenter }}'
        cluster: '{{ vsphere_cluster }}'
        datastore: '{{ vsphere_datastore }}'
        name: '{{ vsphere_vm_name }}'
        disk_provisioning: 'thin'
        validate_certs: 'no'
        networks: "{u'VM Network':u'{{ vsphere_network }}'}"
        power_on: 'yes'
        fail_on_spec_warnings: 'yes'
        wait_for_ip_address: 'yes'
        ovf: '/root/ISO/PA-VM-ESX-8.0.5.ova'
      delegate_to: localhost
```

```
 cat vsphere_creds.yml
# vCenter credentials
vcenter_hostname: 'xxx.xxx.xxx.xxx'
vcenter_username: 'autodeploy@vsphere.local'
vcenter_password: 'xxxxxxxxxxxx'
```
```
# cat vsphere_var.yml
vsphere_datacenter: 'datacenter1'
vsphere_cluster: 'cluster1'
vsphere_datastore: 'management-share'
vsphere_vm_name: 'rhel6-uskar-web'
vsphere_network: 'SDDC-DPortGroup-Mgmt'
vsphere_folder: 'Firewall'
```

To deploy the playbook, run the below command:
```
# ansible-playbook  deploy-ova.yml
```

##### EXPECTED RESULTS
It should be able to attach the NSX based PortGroup and deploy the OVA.

##### ACTUAL RESULTS
OVA deployment failed with the following error:

```
"msg": "Failure validating OVF import spec: Host did not have any virtual network defined.. Host did not have any virtual network defined.. 
```

**Comment by ansibot** [comment]
Files identified in the description:
* [`lib/ansible/modules/cloud/vmware/vmware_deploy_ovf.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/vmware/vmware_deploy_ovf.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @Akasurde @Tomorrow9 @goneri @lparkes @pgbidkar @sivel @warthog9
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@Udayendu, just so you are aware we have a dedicated Working Group for vmware.
You can find other people interested in this in `#ansible-vmware` on Freenode IRC
For more information about communities, meetings and agendas see https://github.com/ansible/community

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by sky-joker** [comment]
Hi, @Udayendu

I tried with NSX-T in my lab environment.

| component | version |
|-----------|---------|
| ESXi      | 6.7.0   |
| vCenter   | 6.7.0   |
| NSX-T     | 2.5.0   |

##### ANSIBLE VERSION

```
ansible --version
ansible 2.9.0rc1
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /root/vmware/ova/venv/lib64/python3.6/site-packages/ansible
  executable location = /root/vmware/ova/venv/bin/ansible
  python version = 3.6.8 (default, Apr 25 2019, 21:02:35) [GCC 4.8.5 20150623 (Red Hat 4.8.5-36)]
```

##### OS/ENVIRONMENT

CentOS 7.5.1804(Core)

```
# pip3 show pyvmomi
Name: pyvmomi
Version: 6.7.3
Summary: VMware vSphere Python SDK
Home-page: https://github.com/vmware/pyvmomi
Author: VMware, Inc.
Author-email: jhu@vmware.com
License: License :: OSI Approved :: Apache Software License
Location: /root/vmware/ova/venv/lib/python3.6/site-packages
Requires: requests, six
Required-by:
```

```
# pip3 show ansible
Name: ansible
Version: 2.9.0rc1
Summary: Radically simple IT automation
Home-page: https://ansible.com/
Author: Ansible, Inc.
Author-email: info@ansible.com
License: GPLv3+
Location: /root/vmware/ova/venv/lib/python3.6/site-packages
Requires: jinja2, PyYAML, cryptography
Required-by:
```

##### STEPS TO REPRODUCE

```
# cat deploy-ova.yml
---
- name: CentOS7 Deployment
  hosts: localhost
  connection: local
  gather_facts: false
  vars_files:
    - vsphere_creds.yml
    - vsphere_var.yml
  tasks:
    - name: VM Deployment from OVA
      vmware_deploy_ovf:
        hostname: '{{ vcenter_hostname }}'
        username: '{{ vcenter_username }}'
        password: '{{ vcenter_password }}'
        datacenter: '{{ vsphere_datacenter }}'
        cluster: '{{ vsphere_cluster }}'
        datastore: '{{ vsphere_datastore }}'
        name: '{{ vsphere_vm_name }}'
        disk_provisioning: 'thin'
        validate_certs: 'no'
        networks: "{u'VM Network':u'{{ vsphere_network }}'}"
        power_on: 'yes'
        fail_on_spec_warnings: 'yes'
        wait_for_ip_address: 'yes'
        ovf: '/root/vmware/ova/CentOS7_TMP.ova'
      delegate_to: localhost
```

```
# cat vsphere_creds.yml
vcenter_hostname: 'xxx.xxx.xxx.xxx'
vcenter_username: administrator@vsphere.local
vcenter_password: 'xxxxxxxxxxxxxx'
```

```
# cat vsphere_var.yml
vsphere_datacenter: 'DC'
vsphere_cluster: 'NSX'
vsphere_datastore: 'VM2'
vsphere_vm_name: 'NewVM'
vsphere_network: 'DPortGroup'
vsphere_folder: 'vm'
```

`DPortGroup` is NSX based PortGroup.

![スクリーンショット 2019-10-15 23 53 16](https://user-images.githubusercontent.com/19516126/66842921-02610380-efa7-11e9-9d51-8473c311190b.png)

##### run playbook

```
# ansible-playbook deploy-ova.yml
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAY [CentOS7 Deployment] *********************************************************************************************************************************************

TASK [VM Deployment from OVA] *****************************************************************************************************************************************
changed: [localhost -> localhost]

PLAY RECAP ************************************************************************************************************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

When I checked the network settings in vCenter, NSX-T PortGroup was set without any problems.

![スクリーンショット 2019-10-16 0 02 41](https://user-images.githubusercontent.com/19516126/66843901-60421b00-efa8-11e9-9c1d-90187dbb0e62.png)

Is the NSX you tried NSX-V? Or is it NSX-T?

**Comment by Akasurde** [comment]
@Udayendu Can you please confirm @sky-joker's findings ? Thanks. 

needs_info

**Comment by sky-joker** [comment]
@Udayendu
I tried the same procedure for NSX-V.

| component | version |
|-----------|---------|
| ESXi      | 6.5.0   |
| vCenter   | 6.5.0   |
| NSX-V     | 6.4.5   |

The target NSX port group are as follows.

![スクリーンショット 2019-10-24 0 53 41](https://user-images.githubusercontent.com/19516126/67412138-b0992880-f5f9-11e9-8aa7-7f4502078cfb.png)

##### STEPS TO REPRODUCE

```
# cat vsphere_var.yml
vsphere_datacenter: 'DC'
vsphere_cluster: 'NSX'
vsphere_datastore: 'VM2'
vsphere_vm_name: 'NewVM'
vsphere_network: 'vxw-dvs-22-virtualwire-1-sid-5000-test001'
vsphere_folder: 'vm'
```

##### run playbook

```
# ansible-playbook deploy-ova.yml
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAY [CentOS7 Deployment] ***********************************************************************************************************************************

TASK [VM Deployment from OVA] *******************************************************************************************************************************
changed: [localhost -> localhost]

PLAY RECAP **************************************************************************************************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

##### result

The port group of NSX-V could be set to VM at the same time as deployment from OVA.

![スクリーンショット 2019-10-24 0 55 14](https://user-images.githubusercontent.com/19516126/67412959-3cf81b00-f5fb-11e9-8b72-6acee915383b.png)

Does the problem event still recur?

**Comment by Udayendu** [comment]
I need to verify again with ansible 2.9.0 GA and then will let you know. We are still on NSX-V.

**Comment by mgsaranraj87** [comment]
Hi All,

Im getting the below error. Im using Distributed Port Group name to assign nic when im trying to deploy VM using vmware_deploy_ovf module.

If Im posting in wrong place , please ignore .

My Environment is

ESXi  - 6.5.0
vCenter -  6.5.0
Ansible .29(latest)

Please find the below output of execution and result.

fatal: [localhost -> localhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "allow_duplicates": true,
            "cluster": "CLU-01",
            "datacenter": "DC",
            "datastore": "DS-CLU-001",
            "deployment_option": null,
            "disk_provisioning": "thin",
            "fail_on_spec_warnings": false,
            "folder": "/DC/vm/LNX/TMP",
            "hostname": "test.lab.com",
            "inject_ovf_env": false,
            "name": "RHEL7-latest-new",
            "networks": {
                "VM Network": "TEST_LAB_3017" **(PortGroup Name)**
            },
            "ovf": "/tmp/RHEL7u6-JUN19.ova",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "port": 443,
            "power_on": false,
            "properties": null,
            "proxy_host": null,
            "proxy_port": null,
            "resource_pool": "Resources",
            "username": "**************",
            "validate_certs": false,
            "wait": true,
            "wait_for_ip_address": false
        }
    },
    "msg": "Failure validating OVF import spec: Host did not have any virtual network defined."
}

Thanks in Advance.

**Comment by Udayendu** [comment]
> Hi All,
> 
> Im getting the below error. Im using Distributed Port Group name to assign nic when im trying to deploy VM using vmware_deploy_ovf module.
> 
> > If Im posting in wrong place , please ignore .
> 
> My Environment is
> 
> ESXi - 6.5.0
> vCenter - 6.5.0
> Ansible 2.9(latest)
> 
> Please find the below output of execution and result.
> 
> fatal: [localhost -> localhost]: FAILED! => {
> "changed": false,
> "invocation": {
> "module_args": {
> "allow_duplicates": true,
> "cluster": "CLU-01",
> "datacenter": "DC",
> "datastore": "DS-CLU-001",
> "deployment_option": null,
> "disk_provisioning": "thin",
> "fail_on_spec_warnings": false,
> "folder": "/DC/vm/LNX/TMP",
> "hostname": "test.lab.com",
> "inject_ovf_env": false,
> "name": "RHEL7-latest-new",
> "networks": {
> "VM Network": "TEST_LAB_3017" **(PortGroup Name)**
> },
> "ovf": "/tmp/RHEL7u6-JUN19.ova",
> "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
> "port": 443,
> "power_on": false,
> "properties": null,
> "proxy_host": null,
> "proxy_port": null,
> "resource_pool": "Resources",
> "username": "**************",
> "validate_certs": false,
> "wait": true,
> "wait_for_ip_address": false
> }
> },
> "msg": "Failure validating OVF import spec: Host did not have any virtual network defined."
> }
> 
> Thanks in Advance.

With dvs it should work. I have tested multiple times. Make sure in your OVA it should have option to accept the dvs based network and you have defined that correctly in you playbooks.

**Comment by mgsaranraj87** [comment]
@Udayendu : Thanks for information. In the networks parameters , im using portgroup name as value and what should be the key value? Will this module accepts "VM Networks" as key value?

It may be silly question , but still im confused what value should i use here

"networks": {
"**<what name should i give>**": "TEST_LAB_3017" (PortGroup Name)
},

Thanks

**Comment by mgsaranraj87** [comment]
I have tried providing Distributed Switch Name  and Portgroup Name as key pair value. Still i get the same error.

**Comment by Udayendu** [comment]
@mgsaranraj87 
Please take a look at the below playbook:

https://github.com/Udayendu/ansible-demo-VmugPNQ/blob/master/roles/vm-deploy/tasks/main.yaml

**Comment by mgsaranraj87** [comment]
@Udayendu : Thanks for sharing the link. 

Im able to create VM using vmware_guest without any issue , even i have tried to manually tried to deploy the VM from OVA file and im able to create the VM successfully.

But when i trying to use vmware_deploy_ovf module , then im getting the below error.
**"msg": "Failure validating OVF import spec: Host did not have any virtual network defined."**

**Comment by sky-joker** [comment]
Hi @mgsaranraj87

Can you check if the target port group can be found using `vmware_dvs_portgroup_info` module?
The following playbook is an example.

```yaml
---
- name: Gather dvs portgroup infomation
  hosts: localhost
  gather_facts: no
  vars:
    hostname: change me to vCenter hostname
    username: administrator@vsphere.local
    password: change me to vCenter user password
    datacenter: change me to datacenter name
  tasks:
    - vmware_dvs_portgroup_info:
        hostname: "{{ hostname }}"
        username: "{{ username }}"
        password: "{{ password }}"
        validate_certs: no
        datacenter: "{{ datacenter }}"
      register: dvpg_info

    - debug: msg="{{ item.portgroup_name }}"
      loop: "{{ dvpg_info.dvs_portgroup_info.DSwitch }}"
      loop_control:
        label: "{{ item.dvswitch_name }}"
```

Please let me know the result of the execution.

**Comment by mgsaranraj87** [comment]
@sky-joker  Thanks for the sharing the playbook. Please find the below result.

(ansible)[root@ansibleserver ansible]# bin/ansible-playbook dvswitch.yml
PLAY [Gather dvs portgroup infomation] ***********************************************************************************************************************

TASK [vmware_dvs_portgroup_info] *****************************************************************************************************************************
ok: [localhost]

TASK [debug] *************************************************************************************************************************************************
ok: [localhost] => {
    "msg": {
        "changed": false,
        "dvs_portgroup_info": {
            "vDS_DATA": [
                {
                    "description": null,
                    "dvswitch_name": "vDS_DATA",
                    "network_policy": {
                        "forged_transmits": false,
                        "mac_changes": false,
                        "promiscuous": false
                    },
                    "num_ports": 82,
                    "port_policy": {
                        "block_override": true,
                        "ipfix_override": false,
                        "live_port_move": false,
                        "network_rp_override": false,
                        "port_config_reset_at_disconnect": true,
                        "security_override": false,
                        "shaping_override": false,
                        "traffic_filter_override": false,
                        "uplink_teaming_override": false,
                        "vendor_config_override": false,
                        "vlan_override": false
                    },
                    "portgroup_name": "TEST_LAB_3064",
                    "teaming_policy": {
                        "inbound_policy": true,
                        "notify_switches": true,
                        "policy": "loadbalance_srcid",
                        "rolling_order": false
                    },
                    "type": "earlyBinding",
                    "vlan_info": {}
                },
                {
                    "description": null,
                    "dvswitch_name": "vDS_DATA",
                    "network_policy": {
                        "forged_transmits": false,
                        "mac_changes": false,
                        "promiscuous": false
                    },
                    "num_ports": 17,
                    "port_policy": {
                        "block_override": true,
                        "ipfix_override": false,
                        "live_port_move": false,
                        "network_rp_override": false,
                        "port_config_reset_at_disconnect": true,
                        "security_override": false,
                        "shaping_override": false,
                        "traffic_filter_override": false,
                        "uplink_teaming_override": false,
                        "vendor_config_override": false,
                        "vlan_override": false
                    },
                    "portgroup_name": "TEST_LAB_3048",
                    "teaming_policy": {
                        "inbound_policy": true,
                        "notify_switches": true,
                        "policy": "loadbalance_srcid",
                        "rolling_order": false
                    },
                    "type": "earlyBinding",
                    "vlan_info": {}
                },
                {
                    "description": null,
                    "dvswitch_name": "vDS_DATA",
                    "network_policy": {
                        "forged_transmits": false,
                        "mac_changes": false,
                        "promiscuous": false
                    },
                    "num_ports": 128,
                    "port_policy": {
                        "block_override": true,
                        "ipfix_override": false,
                        "live_port_move": false,
                        "network_rp_override": false,
                        "port_config_reset_at_disconnect": true,
                        "security_override": false,
                        "shaping_override": false,
                        "traffic_filter_override": false,
                        "uplink_teaming_override": false,
                        "vendor_config_override": false,
                        "vlan_override": false
                    },
                    "portgroup_name": "TEST_LAB_3034",
                    "teaming_policy": {
                        "inbound_policy": true,
                        "notify_switches": true,
                        "policy": "loadbalance_loadbased",
                        "rolling_order": false
                    },
                    "type": "earlyBinding",
                    "vlan_info": {}
                },
                {
                    "description": null,
                    "dvswitch_name": "vDS_DATA",
                    "network_policy": {
                        "forged_transmits": false,
                        "mac_changes": false,
                        "promiscuous": false
                    },
                    "num_ports": 128,
                    "port_policy": {
                        "block_override": true,
                        "ipfix_override": false,
                        "live_port_move": false,
                        "network_rp_override": false,
                        "port_config_reset_at_disconnect": true,
                        "security_override": false,
                        "shaping_override": false,
                        "traffic_filter_override": false,
                        "uplink_teaming_override": false,
                        "vendor_config_override": false,
                        "vlan_override": false
                    },
                    "portgroup_name": "TEST_LAB_3051",
                    "teaming_policy": {
                        "inbound_policy": true,
                        "notify_switches": true,
                        "policy": "loadbalance_loadbased",
                        "rolling_order": false
                    },
                    "type": "earlyBinding",
                    "vlan_info": {}
                },
                {
                    "description": null,
                    "dvswitch_name": "vDS_DATA",
                    "network_policy": {
                        "forged_transmits": false,
                        "mac_changes": false,
                        "promiscuous": false
                    },
                    "num_ports": 128,
                    "port_policy": {
                        "block_override": true,
                        "ipfix_override": false,
                        "live_port_move": false,
                        "network_rp_override": false,
                        "port_config_reset_at_disconnect": true,
                        "security_override": false,
                        "shaping_override": false,
                        "traffic_filter_override": false,
                        "uplink_teaming_override": false,
                        "vendor_config_override": false,
                        "vlan_override": false
                    },
                    "portgroup_name": "TEST_LAB_3054",
                    "teaming_policy": {
                        "inbound_policy": true,
                        "notify_switches": true,
                        "policy": "loadbalance_loadbased",
                        "rolling_order": false
                    },
                    "type": "earlyBinding",
                    "vlan_info": {}
                },
                {
                    "description": null,
                    "dvswitch_name": "vDS_DATA",
                    "network_policy": {
                        "forged_transmits": false,
                        "mac_changes": false,
                        "promiscuous": false
                    },
                    "num_ports": 128,
                    "port_policy": {
                        "block_override": true,
                        "ipfix_override": false,
                        "live_port_move": false,
                        "network_rp_override": false,
                        "port_config_reset_at_disconnect": true,
                        "security_override": false,
                        "shaping_override": false,
                        "traffic_filter_override": false,
                        "uplink_teaming_override": false,
                        "vendor_config_override": false,
                        "vlan_override": false
                    },
                    "portgroup_name": "TEST_LAB_3057",
                    "teaming_policy": {
                        "inbound_policy": true,
                        "notify_switches": true,
                        "policy": "loadbalance_loadbased",
                        "rolling_order": false
                    },
                    "type": "earlyBinding",
                    "vlan_info": {}
                },
                {
                    "description": null,
                    "dvswitch_name": "vDS_DATA",
                    "network_policy": {
                        "forged_transmits": false,
                        "mac_changes": false,
                        "promiscuous": false
                    },
                    "num_ports": 128,
                    "port_policy": {
                        "block_override": true,
                        "ipfix_override": false,
                        "live_port_move": false,
                        "network_rp_override": false,
                        "port_config_reset_at_disconnect": true,
                        "security_override": false,
                        "shaping_override": false,
                        "traffic_filter_override": false,
                        "uplink_teaming_override": false,
                        "vendor_config_override": false,
                        "vlan_override": false
                    },
                    "portgroup_name": "TEST_LAB_3005",
                    "teaming_policy": {
                        "inbound_policy": true,
                        "notify_switches": true,
                        "policy": "loadbalance_loadbased",
                        "rolling_order": false
                    },
                    "type": "earlyBinding",
                    "vlan_info": {}
                },
                {
                    "description": null,
                    "dvswitch_name": "vDS_DATA",
                    "network_policy": {
                        "forged_transmits": false,
                        "mac_changes": false,
                        "promiscuous": false
                    },
                    "num_ports": 140,
                    "port_policy": {
                        "block_override": true,
                        "ipfix_override": false,
                        "live_port_move": false,
                        "network_rp_override": false,
                        "port_config_reset_at_disconnect": true,
                        "security_override": false,
                        "shaping_override": false,
                        "traffic_filter_override": false,
                        "uplink_teaming_override": false,
                        "vendor_config_override": false,
                        "vlan_override": false
                    },
                    "portgroup_name": "TEST_LAB_3003",
                    "teaming_policy": {
                        "inbound_policy": true,
                        "notify_switches": true,
                        "policy": "loadbalance_loadbased",
                        "rolling_order": false
                    },
                    "type": "earlyBinding",
                    "vlan_info": {}
                },
                {
                    "description": null,
                    "dvswitch_name": "vDS_DATA",
                    "network_policy": {
                        "forged_transmits": false,
                        "mac_changes": false,
                        "promiscuous": false
                    },
                    "num_ports": 128,
                    "port_policy": {
                        "block_override": true,
                        "ipfix_override": false,
                        "live_port_move": false,
                        "network_rp_override": false,
                        "port_config_reset_at_disconnect": true,
                        "security_override": false,
                        "shaping_override": false,
                        "traffic_filter_override": false,
                        "uplink_teaming_override": false,
                        "vendor_config_override": false,
                        "vlan_override": false
                    },
                    "portgroup_name": "TEST_LAB_3025",
                    "teaming_policy": {
                        "inbound_policy": true,
                        "notify_switches": true,
                        "policy": "loadbalance_loadbased",
                        "rolling_order": false
                    },
                    "type": "earlyBinding",
                    "vlan_info": {}
                },
                {
                    "description": null,
                    "dvswitch_name": "vDS_DATA",
                    "network_policy": {
                        "forged_transmits": false,
                        "mac_changes": false,
                        "promiscuous": false
                    },
                    "num_ports": 128,
                    "port_policy": {
                        "block_override": true,
                        "ipfix_override": false,
                        "live_port_move": false,
                        "network_rp_override": false,
                        "port_config_reset_at_disconnect": true,
                        "security_override": false,
                        "shaping_override": false,
                        "traffic_filter_override": false,
                        "uplink_teaming_override": false,
                        "vendor_config_override": false,
                        "vlan_override": false
                    },
                    "portgroup_name": "TEST_LAB_3041",
                    "teaming_policy": {
                        "inbound_policy": true,
                        "notify_switches": true,
                        "policy": "loadbalance_loadbased",
                        "rolling_order": false
                    },
                    "type": "earlyBinding",
                    "vlan_info": {}
                },
                {
                    "description": null,
                    "dvswitch_name": "vDS_DATA",
                    "network_policy": {
                        "forged_transmits": false,
                        "mac_changes": false,
                        "promiscuous": false
                    },
                    "num_ports": 256,
                    "port_policy": {
                        "block_override": true,
                        "ipfix_override": false,
                        "live_port_move": false,
                        "network_rp_override": false,
                        "port_config_reset_at_disconnect": true,
                        "security_override": false,
                        "shaping_override": false,
                        "traffic_filter_override": false,
                        "uplink_teaming_override": false,
                        "vendor_config_override": false,
                        "vlan_override": false
                    },
                    "portgroup_name": "TEST_LAB_3039",
                    "teaming_policy": {
                        "inbound_policy": true,
                        "notify_switches": true,
                        "policy": "loadbalance_srcid",
                        "rolling_order": false
                    },
                    "type": "earlyBinding",
                    "vlan_info": {}
                },
                {
                    "description": null,
                    "dvswitch_name": "vDS_DATA",
                    "network_policy": {
                        "forged_transmits": false,
                        "mac_changes": false,
                        "promiscuous": false
                    },
                    "num_ports": 338,
                    "port_policy": {
                        "block_override": true,
                        "ipfix_override": false,
                        "live_port_move": false,
                        "network_rp_override": false,
                        "port_config_reset_at_disconnect": true,
                        "security_override": false,
                        "shaping_override": false,
                        "traffic_filter_override": false,
                        "uplink_teaming_override": false,
                        "vendor_config_override": false,
                        "vlan_override": false
                    },
                    "portgroup_name": "TEST_LAB_3019",
                    "teaming_policy": {
                        "inbound_policy": true,
                        "notify_switches": true,
                        "policy": "loadbalance_srcid",
                        "rolling_order": false
                    },
                    "type": "earlyBinding",
                    "vlan_info": {}
                },
                {
                    "description": null,
                    "dvswitch_name": "vDS_DATA",
                    "network_policy": {
                        "forged_transmits": false,
                        "mac_changes": false,
                        "promiscuous": false
                    },
                    "num_ports": 128,
                    "port_policy": {
                        "block_override": true,
                        "ipfix_override": false,
                        "live_port_move": false,
                        "network_rp_override": false,
                        "port_config_reset_at_disconnect": true,
                        "security_override": false,
                        "shaping_override": false,
                        "traffic_filter_override": false,
                        "uplink_teaming_override": false,
                        "vendor_config_override": false,
                        "vlan_override": false
                    },
                    "portgroup_name": "TEST_INFRA_3035",
                    "teaming_policy": {
                        "inbound_policy": true,
                        "notify_switches": true,
                        "policy": "loadbalance_loadbased",
                        "rolling_order": false
                    },
                    "type": "earlyBinding",
                    "vlan_info": {}
                },
                {
                    "description": null,
                    "dvswitch_name": "vDS_DATA",
                    "network_policy": {
                        "forged_transmits": false,
                        "mac_changes": false,
                        "promiscuous": false
                    },
                    "num_ports": 8,
                    "port_policy": {
                        "block_override": true,
                        "ipfix_override": false,
                        "live_port_move": false,
                        "network_rp_override": false,
                        "port_config_reset_at_disconnect": true,
                        "security_override": false,
                        "shaping_override": false,
                        "traffic_filter_override": false,
                        "uplink_teaming_override": false,
                        "vendor_config_override": false,
                        "vlan_override": false
                    },
                    "portgroup_name": "TEST_INFRA_3059",
                    "teaming_policy": {
                        "inbound_policy": true,
                        "notify_switches": true,
                        "policy": "loadbalance_loadbased",
                        "rolling_order": false
                    },
                    "type": "earlyBinding",
                    "vlan_info": {}
                },
                {
                    "description": null,
                    "dvswitch_name": "vDS_DATA",
                    "network_policy": {
                        "forged_transmits": false,
                        "mac_changes": false,
                        "promiscuous": false
                    },
                    "num_ports": 128,
                    "port_policy": {
                        "block_override": true,
                        "ipfix_override": false,
                        "live_port_move": false,
                        "network_rp_override": false,
                        "port_config_reset_at_disconnect": true,
                        "security_override": false,
                        "shaping_override": false,
                        "traffic_filter_override": false,
                        "uplink_teaming_override": false,
                        "vendor_config_override": false,
                        "vlan_override": false
                    },
                    "portgroup_name": "TEST_INFRA_3026",
                    "teaming_policy": {
                        "inbound_policy": true,
                        "notify_switches": true,
                        "policy": "loadbalance_loadbased",
                        "rolling_order": false
                    },
                    "type": "earlyBinding",
                    "vlan_info": {}
                },
                {
                    "description": null,
                    "dvswitch_name": "vDS_DATA",
                    "network_policy": {
                        "forged_transmits": false,
                        "mac_changes": false,
                        "promiscuous": false
                    },
                    "num_ports": 128,
                    "port_policy": {
                        "block_override": true,
                        "ipfix_override": false,
                        "live_port_move": false,
                        "network_rp_override": false,
                        "port_config_reset_at_disconnect": true,
                        "security_override": false,
                        "shaping_override": false,
                        "traffic_filter_override": false,
                        "uplink_teaming_override": false,
                        "vendor_config_override": false,
                        "vlan_override": false
                    },
                    "portgroup_name": "TEST_LAB_3015",
                    "teaming_policy": {
                        "inbound_policy": true,
                        "notify_switches": true,
                        "policy": "loadbalance_loadbased",
                        "rolling_order": false
                    },
                    "type": "earlyBinding",
                    "vlan_info": {}
                },
                {
                    "description": null,
                    "dvswitch_name": "vDS_DATA",
                    "network_policy": {
                        "forged_transmits": false,
                        "mac_changes": false,
                        "promiscuous": false
                    },
                    "num_ports": 128,
                    "port_policy": {
                        "block_override": true,
                        "ipfix_override": false,
                        "live_port_move": false,
                        "network_rp_override": false,
                        "port_config_reset_at_disconnect": true,
                        "security_override": false,
                        "shaping_override": false,
                        "traffic_filter_override": false,
                        "uplink_teaming_override": false,
                        "vendor_config_override": false,
                        "vlan_override": false
                    },
                    "portgroup_name": "TEST_LAB_3016",
                    "teaming_policy": {
                        "inbound_policy": true,
                        "notify_switches": true,
                        "policy": "loadbalance_loadbased",
                        "rolling_order": false
                    },
                    "type": "earlyBinding",
                    "vlan_info": {}
                },
                {
                    "description": null,
                    "dvswitch_name": "vDS_DATA",
                    "network_policy": {
                        "forged_transmits": false,
                        "mac_changes": false,
                        "promiscuous": false
                    },
                    "num_ports": 60,
                    "port_policy": {
                        "block_override": true,
                        "ipfix_override": false,
                        "live_port_move": false,
                        "network_rp_override": false,
                        "port_config_reset_at_disconnect": true,
                        "security_override": false,
                        "shaping_override": false,
                        "traffic_filter_override": false,
                        "uplink_teaming_override": false,
                        "vendor_config_override": false,
                        "vlan_override": false
                    },
                    "portgroup_name": "TEST_LAB_3018",
                    "teaming_policy": {
                        "inbound_policy": true,
                        "notify_switches": true,
                        "policy": "loadbalance_loadbased",
                        "rolling_order": false
                    },
                    "type": "earlyBinding",
                    "vlan_info": {}
                },
                {
                    "description": null,
                    "dvswitch_name": "vDS_DATA",
                    "network_policy": {
                        "forged_transmits": false,
                        "mac_changes": false,
                        "promiscuous": false
                    },
                    "num_ports": 128,
                    "port_policy": {
                        "block_override": true,
                        "ipfix_override": false,
                        "live_port_move": false,
                        "network_rp_override": false,
                        "port_config_reset_at_disconnect": true,
                        "security_override": false,
                        "shaping_override": false,
                        "traffic_filter_override": false,
                        "uplink_teaming_override": false,
                        "vendor_config_override": false,
                        "vlan_override": false
                    },
                    "portgroup_name": "TEST_LAB_3047",
                    "teaming_policy": {
                        "inbound_policy": true,
                        "notify_switches": true,
                        "policy": "loadbalance_loadbased",
                        "rolling_order": false
                    },
                    "type": "earlyBinding",
                    "vlan_info": {}
                },
                {
                    "description": null,
                    "dvswitch_name": "vDS_DATA",
                    "network_policy": {
                        "forged_transmits": false,
                        "mac_changes": false,
                        "promiscuous": false
                    },
                    "num_ports": 128,
                    "port_policy": {
                        "block_override": true,
                        "ipfix_override": false,
                        "live_port_move": false,
                        "network_rp_override": false,
                        "port_config_reset_at_disconnect": true,
                        "security_override": false,
                        "shaping_override": false,
                        "traffic_filter_override": false,
                        "uplink_teaming_override": false,
                        "vendor_config_override": false,
                        "vlan_override": false
                    },
                    "portgroup_name": "**TEST_LAB_3017**",
                    "teaming_policy": {
                        "inbound_policy": true,
                        "notify_switches": true,
                        "policy": "loadbalance_loadbased",
                        "rolling_order": false
                    },
                    "type": "earlyBinding",
                    "vlan_info": {}
                },
            ],
        },
        "failed": false
    }
}

TASK [debug] *************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"msg": "'dict object' has no attribute 'DSwitch'"}

PLAY RECAP ***************************************************************************************************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0

**Comment by Udayendu** [comment]
@sky-joker 
I tested it with ansible 2.9 and its working perfectly with nsx-v.
So this BZ can be closed.

Thanks alot for you help :)

**Comment by sky-joker** [comment]
Thank you @Udayendu for the confirmed :)

**Comment by sky-joker** [comment]
@mgsaranraj87
Could you please create an issue separately from this issue?

**Comment by mgsaranraj87** [comment]
> @mgsaranraj87
> Could you please create a new issue separately from this issues?

i Have created a new issue. Please check and let me know if you required any details.

https://github.com/ansible/ansible/issues/65409

Thanks
